Proposals v2 iteration2

shamil-gadelshin opened the pull request:
Copied PR to the development branch: original PR and commit history can be found here - 
https://github.com/mnaamani/substrate-runtime-joystream/pull/2

blrhc commented:
@bedeho Can we merge this?

shamil-gadelshin committed:
Add proposals modules

- add iteration 2 codex and engine modules

shamil-gadelshin committed:
Delete modules *.lock files

- delete codex/Cargo.lock
- delete engine/Cargo.lock
- update gitignore

bedeho reviewed (changes requested):
- **Execution model:** The model seems very complex, and its not clear how exactly to naturally use it for different types. I already see that the `proposal_code` is not per say meant to be literal WASM code fragments, and it would be very impractical to expect that sort of heavy setup and handling in a codex module. I think the much simpler approach of passing a dispatchable, like is done in `democracy`, is extremely simple and clean. The codex would simply hold the dispatchables it wanted to run, and guard with root origin check. 

- We did not actually properly define what the side-effect of the text proposal should be. I think we should not confuse the proposal body, with the main proposition itself, as the body will possibly include a broader rationale, links to referecnes etc. The actual text proposal being voted on should be a standalone parameter I think. The side-effect would be that the proposal text is stored as in some state in the codex engine. In the future perhaps it should be stored somewhere else, but for now we may as well put it there, although I do like the idea of the codex being stateless.

- As I point out later in the review, actors should not be identified with accounts, but with some configurable role identifier, at least in the engine. In the runtime, we probably want ot use the membership id for the proposer type, and the council member id for the voter type.

- Just to confirm, is it the case that the codex crate is only consuming anything in the proposals engine crate via crate based dependency management? There is no use of relative paths of modules in the directory tree?

shamil-gadelshin committed:
Fix review comments 1

- introduce ProposalId type
- change BTreeSet to the linked_map

shamil-gadelshin committed:
Fix tests and update codex module

shamil-gadelshin committed:
Embed tally results into the proposal object

shamil-gadelshin committed:
Embed votes into the proposal struct

shamil-gadelshin committed:
Change VoteExistsByProposalByAccount cache

- change VoteExistsByProposalByAccount to proposal_id first key
- clear cache on proposal finalization

shamil-gadelshin committed:
Refactor tally() method

- rename and refactor tally() method in functional style
- introduce internal type FinalizedProposalData

shamil-gadelshin committed:
Codex module build fix

shamil-gadelshin committed:
Introduce proposer_id instead of account_id

shamil-gadelshin committed:
Introduce VoterId type instead of AccountId for voting

shamil-gadelshin committed:
Add TODO:  add maximum allowed active proposals

bedeho reviewed (changes requested):
1. I see that you opt to have mutable variables in a few places. I think we should try to minimize this.

2. Something struck me as I was reviewing this about how we would do slashing: I recall we agreed that there should be a separate quorum and threshold for this action. If this is the case, then we must wait the full voting period before trying to decide whether to accept or slash a proposal, otherwise the engine is sensitive to the accidental order in which people may choose to vote. Please check my logic here.

shamil-gadelshin committed:
Remove redundant check on voting

shamil-gadelshin committed:
Add VotingResults type

Migrating from TallyResults and vote calculation in the end of block to vote calculation on the vote() extrinsic call.

shamil-gadelshin committed:
Remove votes from the Proposal

shamil-gadelshin committed:
Introduce define_proposal_decision_status()

Migrating out from TallyResults

shamil-gadelshin committed:
Remove TallyResults and Vote types

shamil-gadelshin committed:
Add comments and refactor

- refactor map-filter-map pattern to the filter_map
- add comments

shamil-gadelshin committed:
Apply cargo fmt

bedeho reviewed (approved):
Looks good!
